### PR TITLE
[DRAFT] Sketch out how we might support non-mpi and non-pinned apps

### LINKED ISSUE #2728: Design solution to disable profiling
> estimate 1

Done criteria:
- [x] Design discussed with the team
- [x] Scope of work understood

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Fixes #2728

### CODE REVIEW

**Comment by cmcantalupo** [comment]
```
cmcantal@mcfly1:~/Git/geopm/service/tmp (issue-2728)]$ geopmlaunch srun -N1 -n1 --geopm-affinity-disable --geopm-timeout=-1 --geopm-ctl=application sleep 10 
srun: Loaded msrsave restore plugin.
srun: Loaded msrsave restore plugin.
srun: Loaded msrsave restore plugin.
Warning: <geopm> geopmpy.launcher: User did not provide OMP_NUM_THREADS, using 84
Controller launch config: ['srun', '-N', '1', '-n', '1', '--export=LD_PRELOAD=libgeopm.so,ALL', 'geopmctl']
srun: Loaded msrsave restore plugin.
srun: Loaded msrsave restore plugin.
srun: Job 9968 step creation temporarily disabled, retrying (Requested nodes are busy)
srun: Step created for job 9968
cmcantal@mcfly1:~/Git/geopm/service/tmp (issue-2728)]$ ls
geopm.report
cmcantal@mcfly1:~/Git/geopm/service/tmp (issue-2728)]$ cat geopm.report 
GEOPM Version: 2.0.0+dev273g97bd41078
Start Time: Fri Dec 02 18:13:41 2022
Profile: "/home/cmcantal/build/geopm/bin/geopmctl"
Agent: monitor
Policy: {}

Hosts:
  mcfly1:
    Application Totals:
      runtime (s): nan
      count: 0
      geopmctl memory HWM (B): 96876
      geopmctl network BW (B/s): nan

```

**Comment by cmcantalupo** [comment]
```
Note: Google Test filter = ApplicationIOTest.passthrough
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ApplicationIOTest
[ RUN      ] ApplicationIOTest.passthrough

GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: platform_get_profile_pids(@0x1af0c18 "test_profile_name")
          Returns: {}
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#knowing-when-to-expect for details.

GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: platform_get_profile_pids(@0x1af0c18 "test_profile_name")
          Returns: {}
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#knowing-when-to-expect for details.

GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: platform_get_profile_pids(@0x1af0c18 "test_profile_name")
          Returns: {}
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#knowing-when-to-expect for details.

GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: platform_get_profile_pids(@0x1af0c18 "test_profile_name")
          Returns: {}
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#knowing-when-to-expect for details.

GMOCK WARNING:
Uninteresting mock function call - returning default value.
    Function call: platform_get_profile_pids(@0x1af0c18 "test_profile_name")
          Returns: {}
NOTE: You can safely ignore the above warning unless this call should not happen.  Do not suppress it by blindly adding an EXPECT_CALL() if you don't mean to enforce the call.  See https://github.com/google/googletest/blob/master/docs/gmock_cook_book.md#knowing-when-to-expect for details.
unknown file: Failure
C++ exception with description "<geopm> Logic error: ApplicationIOImp::do_shutdown called before connect().: at src/ApplicationIO.cpp:100" thrown in the test body.
[  FAILED  ] ApplicationIOTest.passthrough (5001 ms)
[----------] 1 test from ApplicationIOTest (5001 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (5001 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ApplicationIOTest.passthrough

 1 FAILED TEST
 ```

**Comment by cmcantalupo** [comment]
Latest version of the draft creates more or less correct reports and traces:

```
cmcantal@mcfly5:~/tmp]$ geopmlaunch srun -N1 -n1 --geopm-ctl=application  --geopm-timeout=-1 --geopm-profile=sleep-profile --geopm-trace=geopm.trace sleep 60
cmcantal@mcfly5:~/tmp]$ cat geopm.report 
GEOPM Version: 2.0.0+dev293gfc3c40d1b
Start Time: Tue Dec 13 09:27:43 2022
Profile: "sleep-profile"
Agent: monitor
Policy: {}

Hosts:
  mcfly5:
    Application Totals:
      runtime (s): 58.3707
      count: 0
      sync-runtime (s): 58.3707
      package-energy (J): 6458.28
      dram-energy (J): 693.282
      power (W): 110.643
      frequency (%): 119.428
      frequency (Hz): 2.50798e+09
      uncore-frequency (Hz): 2.35021e+09
      geopmctl memory HWM (B): 105040
      geopmctl network BW (B/s): 0
```

File: service/geopmdpy/service.py [hunk 1]
```diff
@@ -675,6 +675,60 @@ def write_control(self, client_pid, control_name, domain, domain_idx, setting):
...
+
+        """
+        self.open_session(user, client_pid)
+        self._active_sessions.start_profile(client_pid, profile_name)
+
+    def stop_profile(self, client_pid):
+        """Begin profiling a user PID
+
+        Called by a thread to enable profiling as part of a named
+        application.
```
Comment (avilcheslopez) [inline service/geopmdpy/service.py:701]: Looks like this was copied from the previous method.

File: service/geopmdpy/system_files.py [hunk 2]
```diff
@@ -723,6 +729,29 @@ def remove_batch_server(self, client_pid):
             sys.stderr.write(f'Warning: {write_fifo_path} file was left over, deleting it now.\n')
             os.unlink(write_fifo_path)
 
+    def start_profile(self, client_pid, profile_name):
+        self.check_client_active(client_pid, 'start_profile')
+        if profile_name in self._profiles:
+            self._profiles[profile_name].add(client_pid)
+        else:
+            self._profiles[profile_name] = {client_pid}
+        self._sessions[client_pid]['profile_name'] = profile_name
```
Comment (avilcheslopez) [inline service/geopmdpy/system_files.py:738]: Note that the way in which the relationship between pids and profiles is being tracked via the two data structures, can allow a situation where, in one direction a pid can be associated with multiple profile names, but in the other direction it is assumed a pid can only have one profile name. More specifically, the ``_profiles`` data structure will allow adding a pid to different profiles names, while the ``_sessions`` data structure will take the last profile with which ``start_profile()`` was called. For example:
- ``start_profile(629, "profile1")``
- ``start_profile(629, "profile2")``

Afterwards, ``stop_profile()`` will only remove the pid from the pid set corresponding to the last profile with which it was associated.

This is a situation that would not really happen given how the code is set up to call this method (via libgeomload's constructor), but it's still something that should be guarded against as this code cannot make any assumptions about how it's going to be called.

File: service/src/PlatformIO.cpp [hunk 3]
```diff
@@ -1429,4 +1429,5 @@ extern "C" {
     {
         return geopm::platform_io().is_valid_value(value) ? 0 : GEOPM_ERROR_INVALID;
     }
+
```
Comment (avilcheslopez) [inline service/src/PlatformIO.cpp:1432]: Nit: unneeded new line

**Pull request closed without merging**